clang/lib/Sema/SemaExpr.cpp

```diff
@@ -6590,8 +6590,34 @@ ExprResult Sema::BuildCallExpr(Scope *Scope, Expr *Fn, SourceLocation LParenLoc,
     return CallExpr::Create(Context, Fn, ArgExprs, Context.DependentTy,
                             VK_PRValue, RParenLoc, CurFPFeatureOverrides());
   }
-  return BuildResolvedCallExpr(Fn, NDecl, LParenLoc, ArgExprs, RParenLoc,
-                               ExecConfig, IsExecConfig);
+
+  Result = BuildResolvedCallExpr(Fn, NDecl, LParenLoc, ArgExprs, RParenLoc,
+                                 ExecConfig, IsExecConfig);
+
+  if (FunctionDecl *FDecl = dyn_cast_or_null<FunctionDecl>(NDecl);
+      FDecl && FDecl->getBuiltinID() == Builtin::BI__builtin_get_counted_by) {
+    if (const MemberExpr *ME = ArgExprs[0]->getMemberExpr()) {
+      bool IsFlexibleArrayMember = ME->isFlexibleArrayMemberLike(
+          Context, getLangOpts().getStrictFlexArraysLevel());
+
+      if (!ME->HasSideEffects(Context) && IsFlexibleArrayMember &&
```

Comment:
> I don’t think we need to reject this if the expression contains side-effects, but we should definitely warn that it isn’t actually going to be evaluated.

Thinking about this again, it’s actually a bit more complicated than that: this only applies if we return null (we don’t evaluate it at all if the expression names a field that isn’t a FAM guarded by a counted_by attribute). If we don’t return null, then the expression is just evaluated normally.

Also as a side note, Clang is very conservative as to what counts as a ‘side effect’, so even a call to e.g.
```c
S* foo(S* s) { return s; }
```
has side effects, because function calls in general *can* have side effects and we don’t try very hard (or not at all, actually) to figure out whether a language construct that *could* have side effects does or doesn’t have side effects. Iirc `HasSideEffects` only ever returns `false` if it’s something that can never have side effects (like builtin operators).

Reply:
> we don’t evaluate it at all

Actually, we should probably evaluate it regardless; that’s probably a more sensible approach and it means we can do away with this entire side effect discussion entirely. If we end up returning null and the expression *doesn’t* have side effects, the optimiser will just remove it anyway, so there’s no reason to deal w/ that here.